src/java/org/apache/cassandra/db/commitlog/CommitLogReplayer.java

```diff
@@ -148,17 +179,22 @@ else if (end < offset || end > reader.length())
         return end;
     }
 
-    private int getStartOffset(long segmentId, int version)
+    private int getStartOffset(long segmentId, int version, int headerSize)
     {
         if (globalPosition.segment < segmentId)
         {
             if (version >= CommitLogDescriptor.VERSION_21)
-                return CommitLogDescriptor.HEADER_SIZE + CommitLogSegment.SYNC_MARKER_SIZE;
+                return headerSize + CommitLogSegment.SYNC_MARKER_SIZE;
             else
                 return 0;
         }
         else if (globalPosition.segment == segmentId)
-            return globalPosition.position;
+        {
+            if (version >= CommitLogDescriptor.VERSION_22)
```

Comment:
In one case the >= is for VERSION_22 and the other it is for VERSION_21.

In the first case should it return headerSize + CommitLogSegment.SYNC_MARKER_SIZE for version 21? Presumably headerSize has to be 0 in that case for it to have equivalent behavior, but I don't know if that is true.

I am guessing that adding compression changed how filtering is done and records are read and dropped instead of seeking into the correct position in the middle of the file and replaying from there as in pre 2.2. That is why for the segment == segmentId case we return the start offset for the segment and not the offset inside the segment.